### ISSUE TITLE
RFC: Remove per-route sagas

### ISSUE THREAD

**Issue by mxstbr** [react-boilerplate/react-boilerplate#1384]
While our current per-route, codesplit saga setup is awesome (thanks @justingreenberg!) it's sadly let to quite a bit of confusion. (see issues #1277, #1274, #1183, #1137 and many more for examples)

What do we think about removing per-route sagas, adding a global `app/saga/` folder and letting people handle the saga lifecycles themselves however they want to?

This would then allow us to get rid of `react-router-redux`, as the only reason we currently have it is `redux-saga`. That, in turn, would then allow us to remove ImmutableJS by default since the only reason we have it is to show the integration with `react-router-redux`, as it's not that easy. Without that, integrating ImmutableJS is super easy, so there's no need to ship with it by default. (ref #441)

Between getting rid of all the support requests and the much simpler runtime setup, I feel like this would be a very beneficial change. What do you think @justingreenberg @gihrig et al.?

**Comment by ipselon** [comment]
> Between getting rid of all the support requests and the much simpler runtime setup, I feel like this would be a very beneficial change.

I totally agree.

**Comment by gihrig** [comment]
I agree, Anything to streamline and simplify is welcome.

**Comment by justingreenberg** [comment]
@mxstbr i **totally support** this decision :)

**Comment by outdooricon** [comment]
I don't know if other users feel this way but I've always looked up to boilerplate as the "right way" to have my react projects set up. It's not so much of an example for me as it is more of the de facto project and I feel like others look up to it in the same way. If ImmutableJS is removed, then I wonder if others will no longer think it is necessary to use since boilerplate isn't using it., which would a shame considering its optimizations when integrated with reselect and react-redux.

**Comment by AustinMCrane** [comment]
yes, i was just about to ask how to add sagas and reducers to a container that doesnt have a separate route (ie: Navbar)

**Comment by Dattaya** [comment]
Hard to know if it's going to reduce the number of new issues or increase it.

> This would then allow us to get rid of react-router-redux

We can remove `react-router-redux` right now and replace it with a simple history listener:

```js
  browserHistory.listen(location => store.dispatch({ type: 'NAVIGATE', location }));
```

but `react-router-redux` has another benefit—it allows for time travel in `redux-devtools`. I don't know about the average, but I want this library in 99% of my React projects, also it was downloaded 80,374 times in the last week according to npm. Per route sagas is a logical feature to support too, as we have per route reducers. Alternative solution could be improving the docs/extracting some part of the boilerplate into their own packages if possible.

**Comment by smashercosmo** [comment]
> as we have per route reducers
As for me, I would also remove per route reducers :) In my opinion reducers shouldn't be bound to routes, because in real world app there is many-to-many relation between pages and reducers/actions.

**Comment by zsherman** [comment]
I actually really like the pattern that's been set here, it's just that the saga cancellation management results in a pretty confusing/messy experience. What if we simply removed the async loading of sagas? That way they wouldn't be re-injected on every route update and we could dodge all of these issues. I'm sure there's a minor performance hit that would come from loading all of the sagas at once but it would significantly reduce the cognitive load that comes with using this boilerplate. Just a thought...

**Comment by gihrig** [comment]
@zsherman 

> ... a minor performance hit that would come from loading all of the sagas at once...

Daemon sagas, an idea that would allow flexibility to load sagas globally or per-route was discussed in #1315. What are your thoughts on that?

**Comment by zsherman** [comment]
Yeah I could get behind that idea. I would say it adds a little more complexity for newcomers but has the benefit of more flexibility. Would still be curious to see how a larger app would handle potentially hundreds of yielded generator functions at the same time.

**Comment by adelsz** [comment]
We are using this boilerplate for quite some time and I would say that per-route sagas have proved to be very useful for us. 
I think the problem isn't the async injection but the current saga routing/cancellation mechanism, which is confusing and very inflexible.
We have devised a slightly different approach, which is to inject the route sagas on `onEnter` and cancel on `onLeave` route events of our router, which works well and makes the saga hierarchy and lifecycle transparent and clear.
This way we can have global sagas as well as per route (even per child-route) sagas that get canceled/injected only when their level of route path changed. Seems to solve the support issues listed by @mxstbr.
Not sure If I was clear enough, but I can submit a PR for this, let me know if it is needed.

**Comment by gihrig** [comment]
@adelsz Would love to see a PR 😄

**Comment by adelsz** [comment]
Good, so here is a simple, proof of concept PR https://github.com/mxstbr/react-boilerplate/pull/1420.

**Comment by lucidsif** [comment]
I removed redux-sagas completely and use apollo-client instead to manage async side effects. I like react-redux-router because of time-travel and I do like using immutable data structures with re-select. 

Anyways, The Apollo docs mention server-side rendering can be implemented in a react-apollo app using rendering functions built in react-apollo. I do not have any experience with server-side rendering, but I will be exploring this since I need this for a production web application. Any thoughts on replacing redux-saga with Apollo (but still keeping redux, react-router-redux, immutableJS, and reselect) and also doing server-side rendering with react-apollo? We can kill two birds with one stone.
http://dev.apollodata.com/react/server-side-rendering.html

**Comment by dimvic** [comment]
As a newcomer to the boilerplate and react in general I must say that per route sagas also confused me at first, but once I got the hang of them I am a total fanboy, it allows for very nice decoupling and while removing the functionality would not make it difficult to reimplement, I believe it is a very good practice to demonstrate in the boilerplate.

What I would suggest is adding a global sagas file ready to use and an example of a saga that is not cancelled when moving between pages, it took me some time to digest everything and quite a lot of reading to come up with a saga that's loaded in `routes.js` and is not cancelled when switching paths for common operations like animating flash messages, logging in and out, etc.

**Comment by outdooricon** [comment]
Exactly what @dimvic suggests is what I think would be excellent. RBP sets the goal posts of how to excellently design an app and I think showing how to properly handle the sorts of scenarios that you're certain to encounter at some point anyway, is what makes this such an amazing project.

**Comment by advance512** [comment]
@lucidsif 
Do you mind sharing your changes for using apollo-client? I want to do the exact same thing, and keep sagas as a single module (no need for it per component).

**Comment by justingreenberg** [comment]
saga lifecycle helpers are included with updated injectors in #1746—feedback is welcome!

closing issue, but please feel free to continue discussion here
